Src/ILGPU/Runtime/ExchangeBuffer.cs

```diff
@@ -381,4 +458,682 @@ protected override void Dispose(bool disposing)
+        /// </summary>
+        public ArrayView<T, Index2> View { get; }
+
+        /// <summary>
+        /// Returns the length of this buffer in bytes.
+        /// </summary>
+        public Index1 LengthInBytes => Buffer.LengthInBytes;
+
+        /// <summary>
+        /// Returns the extent of this buffer.
+        /// </summary>
+        public Index2 Extent { get; }
+
+        /// <summary>
+        /// Returns an array view to the CPU part of this buffer.
+        /// </summary>
+        public ArrayView<T, Index2> CPUView { get; }
+
+        /// <summary>
+        /// Gets/sets the element specified by the index
+        /// </summary>
+        /// <param name="index2">The index to access</param>
+        /// <returns>The specified element</returns>
+        /// <remarks>
+        /// Uses the CPU View (<see cref="CPUView"/>), not the GPU View (<see cref="View"/>). <br></br>
+        /// You will need to copy again if you want use the modified version in a kernel.
+        /// </remarks>
+        public T this[Index2 index2]
+        {
+            get => CPUView[index2];
+
+            set => CPUView[index2] = value;
+        }
+
+        #endregion
+
+        #region Methods
+
+        /// <summary>
+        /// Sets the contents of the current buffer to zero.
+        /// </summary>
+        /// <param name="stream">The used accelerator stream.</param>
+        public override void MemSetToZero(AcceleratorStream stream) =>
+            Buffer.MemSetToZero(stream);
+
+        /// <summary>
+        /// Copies the current contents into a new byte array.
+        /// </summary>
+        /// <param name="stream">The used accelerator stream.</param>
+        /// <returns>A new array holding the requested contents.</returns>
+        public override byte[] GetAsRawArray(AcceleratorStream stream) =>
+            Buffer.GetAsRawArray(stream);
+
+        /// <summary>
+        /// Copies the current contents into a new byte array.
+        /// </summary>
+        /// <param name="stream">The used accelerator stream.</param>
+        /// <param name="byteOffset">The offset in bytes.</param>
+        /// <param name="byteExtent">The extent in bytes (number of elements).</param>
+        /// <returns>A new array holding the requested contents.</returns>
+        protected internal override ArraySegment<byte> GetAsRawArray(
+            AcceleratorStream stream,
+            Index1 byteOffset,
+            Index1 byteExtent) =>
+            Buffer.GetAsRawArray(stream, byteOffset, byteExtent);
+
+        /// <summary>
+        /// Copies the current contents into a new array.
+        /// </summary>
+        /// <param name="stream">The used accelerator stream.</param>
+        /// <returns>A new array holding the requested contents.</returns>
+        public unsafe T[] GetAsArray(AcceleratorStream stream)
+        {
+            CopyFromAccelerator(stream);
+            stream.Synchronize();
+
+            var data = new T[Length];
```

Comment:
It feels like this functionality already exists in the code base. Maybe you could find another way of expressing this without these code lines?

Reply:
You're right, the functionality is already there. I fixed it.